src/datasource.ts

```diff
@@ -17,4 +38,157 @@ export class DataSource extends DataSourceWithBackend<RedshiftQuery, RedshiftDat
 
   applyTemplateVariables = (query: RedshiftQuery, scopedVars: ScopedVars) =>
     applySQLTemplateVariables(query, scopedVars, getTemplateSrv);
+
+  query(request: DataQueryRequest<RedshiftQuery>): Observable<DataQueryResponse> {
+    if (config.featureToggles.redshiftAsyncQueryDataSupport) {
+      const targets = request.targets.filter(this.filterQuery);
+      if (!targets.length) {
+        return of({ data: [] });
+      }
+      const all: Array<Observable<DataQueryResponse>> = [];
+      for (let target of targets) {
+        if (target.hide) {
+          continue;
+        }
+        all.push(this.doSingle(target, request));
+      }
+      return merge(...all);
+    } else {
+      return super.query(request);
+    }
+  }
+
+  storeQuery(target: RedshiftQuery, queryInfo: RedshiftRunningQueryInfo) {
+    const key = JSON.stringify(target);
+    const existingQueryInfo = this.runningQueries[key] || {};
+    this.runningQueries[key] = { ...existingQueryInfo, ...queryInfo };
+  }
+
+  getQuery(target: RedshiftQuery) {
+    const key = JSON.stringify(target);
+    return this.runningQueries[key] || {};
+  }
+
+  removeQuery(target: RedshiftQuery) {
+    const key = JSON.stringify(target);
+    delete this.runningQueries[key];
+  }
+
+  doSingle(target: RedshiftQuery, request: DataQueryRequest<RedshiftQuery>): Observable<DataQueryResponse> {
+    let queryID: string | undefined = undefined;
+    let status: string | undefined = undefined;
+    let allData: DataFrame[] = [];
+
+    return getRequestLooper(
+      { ...request, targets: [target], requestId: `aws_ts_${requestCounter++}` },
+      {
+        /**
+         * Additional query to execute if the current query is still in a running state
+         */
+        getNextQuery: (rsp: DataQueryResponse) => {
+          if (rsp.data?.length) {
+            const first: DataFrame = rsp.data[0];
+            const meta = first.meta?.custom;
+
+            if (isRedshiftCustomMeta(meta) && isRunning(meta.status)) {
+              queryID = meta.queryID;
+              status = meta.status;
+              this.storeQuery(target, { queryID });
+              return { ...target, queryID };
+            }
+          }
+
+          this.removeQuery(target);
+          return undefined;
+        },
+
+        /**
+         * The original request
+         */
+        query: (request: DataQueryRequest<RedshiftQuery>) => {
+          const { range, targets, requestId, intervalMs, maxDataPoints } = request;
+          const [_query] = targets;
+          const query: RedshiftQuery = {
+            ..._query,
+            ...(config.featureToggles.redshiftAsyncQueryDataSupport ? { meta: { queryFlow: 'async' } } : {}),
+          };
+
+          const data = {
+            queries: [
+              {
+                ...query,
+                intervalMs,
+                maxDataPoints,
+                datasource: this?.getRef(),
+                datasourceId: this.id,
```

Comment:
`this?.getRef()` is intentional in case someone tries to run this in a Grafana version pre 8.3.x which I think is before `getRef` was added.

`datasourceId` is also added for backwards compat.

Reply:
i feel like it might be worth adding this comment to the codebase, i could see someone accidentally removing it